<a href="https://colab.research.google.com/github/Isha-Sisodia/DSA-Practice/blob/main/Strings/6.%20Count%20and%20Say.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**🔁 How do we get the next term?**

We apply Run-Length Encoding (RLE) on the previous string.

**RLE means:**

Read the string left to right

Count consecutive same digits

Say count + digit

🧠 Let’s understand with examples
Step-by-step build
n = 1
"1"
n = 2 → read "1"
One 1 → "11"
n = 3 → read "11"
Two 1s → "21"
n = 4 → read "21"
One 2 → "12"
One 1 → "11"
Result → "1211"
So:

1 → 11 → 21 → 1211 → ...

**before start question 6 ...i think we should practice some questions**
### **Practice Question 1**
Q1️⃣: Count consecutive characters (NO loop inside loop)

Input: "aaabbc"
Output: "a3b2c1"

👉 Write a function that:

counts consecutive characters

returns character followed by count

In [ ]:
# i understood the logic of this question after dry run this code
def func(s):
    count = 1
    result = ""

    for i in range(1, len(s)):
        if s[i] == s[i-1]:
            count += 1
        else:
          result += s[i-1] + str(count)
          count = 1

    result += s[-1] + str(count)
    return result
s = "aaadeecc"
print(func(s))

a3d1e2c2


In [ ]:
# This logic is mine after understanding the above logic...
def check(s):
  count = 1
  result = ""
  for i in range(0,len(s)-1):
    if s[i] == s[i+1]:
      count += 1
    else:
      result += s[i] + str(count)
      count = 1
  result += s[-1] + str(count)
  return result
s = "aaadeecc"
print(check(s))

a3d1e2c2


🔹 PRACTICE 2 (Count-and-Say Single Step)
### **Practice Question 2**
Q2️⃣: Perform one step of count-and-say (Run-Length Encoding (RLE))

Input: "1211"
Output: "111221"

Explanation:

1 → one 1

2 → one 2

11 → two 1s

In [2]:
def say_once(s):
  count = 1
  result = ""
  for i in range(0, len(s)-1):
    if s[i] == s[i+1]:
      count += 1
    else:
      result += str(count) + s[i]
      count = 1
  result += str(count) + s[-1]
  return result
print(say_once("11211"))
print(say_once("3322251"))


211221
23321511


## 🔹 **PRACTICE Question 3** (Main Problem – Simplified)

Q3️⃣: Count-and-say till n = 5

Start with "1"

Apply say_once repeatedly.

In [ ]:
s = "1"
n = 5
print(s)
for i in range(1,n):
    s = say_once(s)
    print(s)

1
11
21
1211
111221


# **Sheet Question**

The count-and-say sequence is a sequence of digit strings defined by the recursive formula:

countAndSay(1) = "1"
countAndSay(n) is the run-length encoding of countAndSay(n - 1).
Run-length encoding (RLE) is a string compression method that works by replacing consecutive identical characters (repeated 2 or more times) with the concatenation of the character and the number marking the count of the characters (length of the run). For example, to compress the string "3322251" we replace "33" with "23", replace "222" with "32", replace "5" with "15" and replace "1" with "11". Thus the compressed string becomes "23321511".

Given a positive integer n, return the nth element of the count-and-say sequence.



Example 1:

Input: n = 4

Output: "1211"

Explanation:

countAndSay(1) = "1"                     that means we just have to run the loop 3 times because first mein toh "1" hi aayega...which is our base case
countAndSay(2) = RLE of "1" = "11"
countAndSay(3) = RLE of "11" = "21"
countAndSay(4) = RLE of "21" = "1211"
Example 2:

Input: n = 1

Output: "1"

Explanation:

This is the base case.



Constraints:

1 <= n <= 30

In [ ]:
# https://leetcode.com/problems/count-and-say/
class Solution(object):
    def countAndSay(self, n):
        s = "1"
        for i in range(n-1):
            result = ""
            count = 1
            for i in range(0,len(s)-1):
                if s[i] == s[i+1]:
                    count += 1
                else:
                    result += str(count) + s[i]
                    count = 1
            result += str(count) + s[-1]
            s = result
        return s

